src/VisualStudio/Core/Def/Implementation/EditAndContinue/VisualStudioActiveStatementProvider.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.Debugger;
+using Microsoft.VisualStudio.Debugger.Clr;
+using Microsoft.VisualStudio.Debugger.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.EditAndContinue
+{
+    [Export(typeof(IActiveStatementProvider)), Shared]
+    internal sealed partial class VisualStudioActiveStatementProvider : IActiveStatementProvider
+    {
+        public Task<ImmutableArray<ActiveStatementDebugInfo>> GetActiveStatementsAsync(CancellationToken cancellationToken)
+        {
+            DkmComponentManager.InitializeThread(DebuggerComponentIds.ManagedEditAndContinueService);
+            try
+            {
+                // TODO: report errors
+                // TODO: return empty outside of debug session.
+                // https://github.com/dotnet/roslyn/issues/24325
+
+                var workList = DkmWorkList.Create(CompletionRoutine: null);
+                var completion = new TaskCompletionSource<ImmutableArray<ActiveStatementDebugInfo>>();
+                var builders = default(ArrayBuilder<ArrayBuilder<ActiveStatementDebugInfo>>);
+                int pendingRuntimes = 0;
+                int runtimeCount = 0;
+
+                foreach (var process in DkmProcess.GetProcesses())
+                {
+                    foreach (var runtimeInstance in process.GetRuntimeInstances())
+                    {
+                        if (runtimeInstance.TagValue == DkmRuntimeInstance.Tag.ClrRuntimeInstance)
+                        {
+                            var clrRuntimeInstance = (DkmClrRuntimeInstance)runtimeInstance;
+
+                            int runtimeIndex = runtimeCount;
+                            clrRuntimeInstance.GetActiveStatements(workList, activeStatementsResult =>
+                            {
+                                if (cancellationToken.IsCancellationRequested)
+                                {
+                                    workList.Cancel();
+                                    completion.SetCanceled();
```

Comment:
Oh this is silly: there's no SetCanceled that takes a cancellation token. Le sigh. Can you call TrySetCanceled(cancellationToken) to make the TPL happy in some obscure cases and regret the lack of better API here?

(the TPL sometimes cares that a task is cancelled with a certain token, which is why this matters.)